balance_strategy.go

```diff
@@ -127,3 +155,899 @@ func balanceStrategyHashValue(vv ...string) uint32 {
+	}
+}
+
+// Calculate the balance score of the given assignment, as the sum of assigned partitions size difference of all consumer pairs.
+// A perfectly balanced assignment (with all consumers getting the same number of partitions) has a balance score of 0.
+// Lower balance score indicates a more balanced assignment.
+func getBalanceScore(assignment map[string][]topicPartitionAssignment) int {
+	consumer2AssignmentSize := make(map[string]int, len(assignment))
+	for memberID, partitions := range assignment {
+		consumer2AssignmentSize[memberID] = len(partitions)
+	}
+
+	var score float64
+	for memberID, consumerAssignmentSize := range consumer2AssignmentSize {
+		delete(consumer2AssignmentSize, memberID)
+		for _, otherConsumerAssignmentSize := range consumer2AssignmentSize {
+			score += math.Abs(float64(consumerAssignmentSize - otherConsumerAssignmentSize))
+		}
+	}
+	return int(score)
+}
+
+// Determine whether the current assignment plan is balanced.
+func isBalanced(currentAssignment map[string][]topicPartitionAssignment, sortedCurrentSubscriptions []string, allSubscriptions map[string][]topicPartitionAssignment) bool {
+	sortedCurrentSubscriptions = sortMemberIDsByPartitionAssignments(currentAssignment)
+	min := len(currentAssignment[sortedCurrentSubscriptions[0]])
+	max := len(currentAssignment[sortedCurrentSubscriptions[len(sortedCurrentSubscriptions)-1]])
+	if min >= max-1 {
+		// if minimum and maximum numbers of partitions assigned to consumers differ by at most one return true
+		return true
+	}
+
+	// create a mapping from partitions to the consumer assigned to them
+	allPartitions := make(map[topicPartitionAssignment]string)
+	for memberID, partitions := range currentAssignment {
+		for _, partition := range partitions {
+			if _, exists := allPartitions[partition]; exists {
+				Logger.Printf("Topic %s Partition %d is assigned more than one consumer", partition.Topic, partition.Partition)
+			}
+			allPartitions[partition] = memberID
+		}
+	}
+
+	// for each consumer that does not have all the topic partitions it can get make sure none of the topic partitions it
+	// could but did not get cannot be moved to it (because that would break the balance)
+	for _, memberID := range sortedCurrentSubscriptions {
+		consumerPartitions := currentAssignment[memberID]
+		consumerPartitionCount := len(consumerPartitions)
+
+		// skip if this consumer already has all the topic partitions it can get
+		if consumerPartitionCount == len(allSubscriptions[memberID]) {
+			continue
+		}
+
+		// otherwise make sure it cannot get any more
+		potentialTopicPartitions := allSubscriptions[memberID]
+		for _, partition := range potentialTopicPartitions {
+			if !memberAssignmentsIncludeTopicPartition(currentAssignment[memberID], partition) {
+				otherConsumer := allPartitions[partition]
+				otherConsumerPartitionCount := len(currentAssignment[otherConsumer])
+				if consumerPartitionCount < otherConsumerPartitionCount {
```

Comment:
I'm fairly certain the Java code here _should_ check `consumerPartitionCount < otherConsumerPartitionCount - 1`. A delta of one is fine, as was checked at the beginning of this function. Otherwise, moving the partition from the other to this one would create the same balance in reverse.

But,

Reply:
Well, I guess it's fine. Even if this results in a partition moving from a member with 3 to a member with 2, the worst that can happen is the loop in `performReassignments` will continue unnecessarily.

If I'm reading it correctly, the one-off imbalance could cause performReassignments to loop across all reassignable partitions no matter what. If a plan has the two fewest partition members A and B that can have partitions 1, 2, and 3, if they look like

A: 1,
B: 2, 3

As a further constraint, they can take no more partitions.

That condition would cause the first conditional in `performReassignments` to always return true.

But, the loop stops if a modification is made, so that's good.

The more I think about it, I think just checking `consumerPartitionCount ... other - 1` would be flawed as well, and in a potentially worse way.

I think what would be an accurate check here would be, in addition to the current check (not my modification), to also compare the sum of both member's partitions against the union of the potential partitions they could consume from.

This would return false on what I suggested above, but would return true if B could consume from a fourth partition but isn't yet.

But, that union suggestion may be overkill, and if this is going for parity with Java, the Java stuff doesn't do that.

Reply:
FWIW I tested this union idea in this branch; it kicks in three times during the tests but doesn't change the test results. I do notice that some code in the partitionMovements methods below does some back-and-forth checking. Worst case, Java does a move-once type of scenario when it doesn't need to, and even that may be mitigated by some other duct tape.